test/integration/multicluster/target3/testdata/nginx-statefulset.yml

```diff
@@ -0,0 +1,35 @@
+apiVersion: v1
+kind: Service
+metadata:
+  name: nginx-svc
+  namespace: default
+spec:
+  clusterIP: None
+  clusterIPs:
+  - None
+  selector:
+    app: nginx-set
+  sessionAffinity: None
+  type: ClusterIP
+  ports:
+  - protocol: TCP
+    port: 80
+---
+apiVersion: apps/v1
+kind: StatefulSet
+metadata:
+  name: nginx-set
+spec:
+  serviceName: nginx-svc
+  selector:
+    matchLabels:
+      app: nginx-set
+  replicas: 3
```

Comment:
fwiw: one thing I'd like to test with headless services is scaling. Here's a suggestion, what if we first switch this replica count to 1 (or 2) -- with one replica, the test will start up a bit faster (StatefulSets start in order, unlike pods there's a bit of a dependency between replicas so it takes longer for them to come up) -- we can then scale up to 2 replicas and test that additional services have been mirrored in the source cluster. Luckily for us, StatefulSets have predictable names so we can easily assert for example an Endpoints object has the right state after scaling up.

Not a strict necessity, but it would definitely be good to test.

Reply:
Not sure I understand, which scenario are you talking about here?
1. First write the integration test with 1 replica; after we do that we can try with 2 and that it still works.
2. In the same manifest, make sure that the 2nd replica starts setting up after the 1st is done. (If this is the one, don't tell me how to do it, I'd like to figure it out myself first)

Or is it something else?

Reply:
Okay, for context, remember that a headless service that is mirrored (also known as a headless mirror) will be a "root" service. It will be a root because its endpoints object will contain addresses that belong to other services -- services that we create based on the pods in the StatefulSet (we call these endpoint mirrors).

What I'd like to test is the logic in the service-mirror but in a less controlled environment, i.e in the k3d clusters. We can apply the manifest with one replica, and that's part of our initial test ("can we deploy a StatefulSet with a headless service in a target cluster and see it mirrored in a source cluster?" is the first thing we should be testing here). A good follow-up thing to do would be to scale our StatefulSet up in the target cluster, rather than re-apply a manifest (with 2 replicas instead of 1) I'd like to just scale up the object itself (as an operator would in k8s through `kubectl` or whatever). We know ahead of time what the behaviour should be when a StatefulSet is scaled up (I'll leave these to you, there are a few, but for example we expect another `clusterIP` endpoint mirror to be created in the source cluster). So, this scaling test would just verify the assumptions that we have about the system in a real world environment. Hope this makes a bit more sense.

This is what I'd like to test, it's more in line with (1).